src/endian/detail/stream.hpp

```diff
@@ -16,26 +16,22 @@ namespace endian
 namespace detail
 {
 
+using const_stream = std::true_type;
+using non_const_stream = std::false_type;
+
 /// @brief Base-class for the endian stream reader and writer.
-template <typename DataPointerType>
+template <typename use_const = detail::non_const_stream>
```

Comment:
Generally I like this approach. It makes it clear that the choice is binary.
Typically we use CamelCase for template arguments so use_const should be UseConst, also I think it would be better to not have the default. Maybe we should also have a check that the provided template argument is a true/false_type.